vllm/entrypoints/openai/tool_parsers/mistral_tool_parser.py

```diff
@@ -0,0 +1,304 @@
+import json
+import re
+from typing import Dict, List, Optional, Union
+
+import partial_json_parser
+from partial_json_parser.core.options import Allow
+from transformers import (AutoTokenizer, PreTrainedTokenizer,
+                          PreTrainedTokenizerFast)
+
+from vllm.entrypoints.openai.protocol import (DeltaFunctionCall, DeltaMessage,
+                                              DeltaToolCall,
+                                              ExtractedToolCallInformation,
+                                              FunctionCall,
+                                              InitialDeltaToolCall, ToolCall)
+from vllm.entrypoints.openai.tool_parsers.abstract_tool_parser import (
+    ToolParser)
+from vllm.entrypoints.openai.tool_parsers.utils import (
+    extract_intermediate_diff)
+from vllm.logger import init_logger
+
+logger = init_logger(__name__)
+
+
+class MistralToolParser(ToolParser):
+    """
+    Tool call parser for Mistral 7B Instruct v0.3, intended for use with the
+    examples/tool_chat_template_mistral.jinja template.
+
+    Used when --enable-auto-tool-choice --tool-call-parser gmistral are all set
+    """
+
+    # the bot_token is the token indicating tool call(s) follow. Tokens before
+    # this token will be parsed as content; and
+    # if not present, the entire response will be parsed as text content.
+    bot_token: str = "[TOOL_CALLS]"  # string literal
+    bot_token_id: int = 5  # token ID thereof from the models" tokenizer
```

Comment:
weird commend and this value seems very hardcoded. is there a way you could use the model's tokenizer to produce this token id or validate this?

Reply:
I refactored this to use non-static methods here so that the instances can access the tokenizer - c87a81f2
The token ID for the BOT token is looked up from the tokenizer based on the string for it, but unfortunately there's not a universal standard on how to dictate which tokens in the tokenizer indicate the presence of tool calls, so the tokenizer has to know which token to look for somehow - but, we can retrieve the token ID for it given the string.